client.go

```diff
@@ -302,15 +302,25 @@ func (c *Client) CreateDataset(options *CreateDatasetOptions) (datasetState *Dat
 	tickChan := time.NewTicker(tickerWaitingForVolume).C
 
 	for {
-		if s, err := c.GetDatasetState(p.DatasetID); err == nil {
+		var strErrDel string
+		s, err := c.GetDatasetState(p.DatasetID)
+		if err == nil {
 			return s, nil
 		} else if err != errStateNotFound {
-			return nil, err
+			errDel := c.DeleteDataset(p.DatasetID)
+			if errDel != nil {
+				strErrDel = fmt.Sprintf(", deletion of dataset failed with %s", errDel)
+			}
+			return nil, fmt.Errorf("Flocker API error during dataset creation (datasetID %s): %s%s", p.DatasetID, err, strErrDel)
 		}
 
 		select {
 		case <-timeoutChan:
-			return nil, err
```

Comment:
In https://github.com/ClusterHQ/flocker-go/pull/12#issuecomment-247573709 @simonswine wrote:
> The biggest problem/bug right now for k8s dynamic provisioning is that if CreateDataset runs into a timeout, it will return nil, nil which crashes Kubernetes Controller Manager

Ah, so this is the important bit. It's returning `err` from the *successful* API call above.